### ISSUE TITLE
Help for native Ads

### ISSUE THREAD

**Issue by FVivolo** [mopub/mopub-ios-sdk#266]
One line summary of the issue here

- [ ] I am submitting a bug report for existing functionality
- [ X] I visited https://developers.mopub.com/ and found no answer
- [X ] I checked https://twittercommunity.com/c/advertiser-api/mopub to make sure that this issue has not already been filed
- [ ] I checked to make sure that this issue has not already been filed

#### MoPub SDK Version:
5.6.0
#### Device model and OS Version:
Ipad Air 2
#### Xcode Version:
10.2.1
#### Ad Unit IDs used in reproducing the issue:
The one that is used for native ad/video  in MoPub sample

#### Expected behavior:
Please list all relevant steps to reproduce the observed behavior.
i am executing your same code for display native ads/video as is mentioned in the MoPub example but only get the view without icon or ads and mentioning success, is this correct, what i am missing, i can not use xib as mentioned in your example and only creating the adviewcontainer and used inside my own view controller, i will appreciate all your help, i will attach two images of what i get.

#### Observed behavior:
As concisely as possible, describe the observed behavior.

#### Evidence:
Device log files, Network log file, etc.
![Screen Shot 2019-04-24 at 9 59 21 AM](https://user-images.githubusercontent.com/4428409/56972873-cb527f00-6b39-11e9-8b83-a42ddc229e60.png)
![Screen Shot 2019-04-30 at 11 20 33 AM](https://user-images.githubusercontent.com/4428409/56972988-fc32b400-6b39-11e9-899e-fe2f11d0a560.png)

**Comment by chauduyphanvu1** [comment]
@FVivolo, since you are using our sample ad unit ID, all ad assets should be returned. That'd mean this is an implementation issue. 

Did you create the two `UIImageView`s for the [icon image](https://github.com/mopub/mopub-ios-sdk/blob/1f4b83e7fefedc9fea517df5448856c63e4f16d9/MoPubSampleApp/Views/MPStaticNativeAdView.m#L29-L30) and the [main image](https://github.com/mopub/mopub-ios-sdk/blob/1f4b83e7fefedc9fea517df5448856c63e4f16d9/MoPubSampleApp/Views/MPStaticNativeAdView.m#L32-L35) assets? If yes, are they present in your [view hierarchy](https://developer.apple.com/library/archive/documentation/ToolsLanguages/Conceptual/Xcode_Overview/ExaminingtheViewHierarchy.html)? Then, make sure to pass your ad view to the `MPStaticNativeAdRendererSettings`, like [here](https://github.com/mopub/mopub-ios-sdk/blob/1f4b83e7fefedc9fea517df5448856c63e4f16d9/MoPubSampleApp/Controllers/MPNativeAdDetailViewController.m#L82-L83).

Vu Chau
MoPub SDK Team

**Comment by FVivolo** [comment]
mm i am using both of your ViewClass as you do in your example and i get the call to 

- (UIImageView *)nativeIconImageView
{
    return self.iconImageView;
}

- (UIImageView *)nativeMainImageView
{
    return self.mainImageView;
}

for video i get this

- (UIView *)nativeVideoView
{
    return self.videoView;
}
in both the ad class and the video class

also using the same loadAd code partially modify

    [self clearAd];
    
    MPStaticNativeAdRendererSettings *settings = [[MPStaticNativeAdRendererSettings alloc] init];
    settings.renderingViewClass = [FSMPNativeAdView class];
    
    MPNativeAdRendererConfiguration *config = [MPStaticNativeAdRenderer rendererConfigurationWithRendererSettings:settings];
    NSMutableArray * configurations = [NSMutableArray arrayWithObject:config];
    
    MOPUBNativeVideoAdRendererSettings *nativeVideoAdSettings = [[MOPUBNativeVideoAdRendererSettings alloc] init];
    nativeVideoAdSettings.renderingViewClass = [FSMPNativeVideoView class];
    nativeVideoAdSettings.viewSizeHandler = ^(CGFloat maximumWidth) {
        return CGSizeMake(maximumWidth, 312.0f);
    };
    
    MPNativeAdRendererConfiguration *nativeVideoConfig = [MOPUBNativeVideoAdRenderer rendererConfigurationWithRendererSettings:nativeVideoAdSettings];
    [configurations addObject:nativeVideoConfig];
    
    MPNativeAdRequest *adRequest1 = [MPNativeAdRequest requestWithAdUnitIdentifier:NativeAdID rendererConfigurations:configurations];
    
    MPNativeAdRequestTargeting *targeting = [MPNativeAdRequestTargeting targeting];
    targeting.desiredAssets = [NSSet setWithObjects:kAdTitleKey, kAdTextKey, kAdCTATextKey, kAdIconImageKey, kAdMainImageKey, kAdStarRatingKey, nil];
    adRequest1.targeting = targeting;
    [adRequest1 startWithCompletionHandler:^(MPNativeAdRequest *request, MPNativeAd *response, NSError *error) {
        isNativeShow = NO;
        if (error) {
            NSLog(@"Native Ad Error> %@", error);
            [self pushError:(int)error.code];
        } else {
            NSLog(@"Received Native Video");
            nativeAd = response;
            nativeAd.delegate = self;
            [[adViewContainer subviews] makeObjectsPerformSelector:@selector(removeFromSuperview)];
            UIView *adView = [nativeAd retrieveAdViewWithError:nil];
            [adViewContainer addSubview:adView];
            adView.frame = adViewContainer.bounds;
            [mview addSubview:adViewContainer];
            adViewContainer.accessibilityLabel = kNativeAdDefaultActionViewKey;
            isNativeShow = YES;
            ///// internal code [ho pushEvent:CNDONNATIVELOAD withParam:0];
        }
    }];

and the adViewContainer created somewhere in my code as

   adViewContainer = [[UIView alloc] init];
    /// my mainview ----> mview = ho->hoAdRunHeader->rhApp->mainView;
 and sized manually

all the flow seems to be the same as your example but no ads or video

**Comment by FVivolo** [comment]
Any Help about this, can you provide other example related to native ads which not involve xib? as you may see teh hierarchy seems to be like teh one in your example and still does not show.

**Comment by chauduyphanvu1** [comment]
Your code looks OK to me. Could you check Xcode's console log to see if there are any relevant messages that might explain images failing to load? Are you also using any proxy that might be causing images to fail to load in general? You can run our iOS sample app in parallel and see if the images are displayed.

Feel free to attach your app project and we can further investigate!

**Comment by FVivolo** [comment]
2019-05-01 10:17:01.571175-0400 RuntimeIPhone[479:99021] 
	[MoPub][MPNativeAdRequest][76a3fefaced247959582d2d2df6f4757] Attempting to load ad
2019-05-01 10:17:01.665795-0400 RuntimeIPhone[479:99119] 
	[MoPub][MPNativeAdRequest] Received data from MoPub to construct native ad.
2019-05-01 10:17:01.667713-0400 RuntimeIPhone[479:99003] 
	[MoPub][MPNativeAdRequest] Looking for custom event class named MPMoPubNativeCustomEvent.
2019-05-01 10:17:01.849869-0400 RuntimeIPhone[479:98895] Received Native Video
2019-05-01 10:17:01.849954-0400 RuntimeIPhone[479:99119] 
	[MoPub][MPNativeAdRequest][76a3fefaced247959582d2d2df6f4757] Ad loaded


also is not trigger is the delegate.

my test project have some proprietary code can i send you a link via email?

**Comment by chauduyphanvu1** [comment]
Could you remove the proprietary code and keep _only_ the ad/MoPub code? Could you upload it to Google Drive, share a link here, and grant access to me when I request it? Thank you.

**Comment by FVivolo** [comment]
https://drive.google.com/drive/folders/1Adb9qCBRLmC59oHCuPtZ-16WIKwLo0dS?usp=sharing

**Comment by FVivolo** [comment]
the code is inside two files, but inside you will see several implementation/interface this is due the way to pack this code as object for clickteam runtime

**Comment by chauduyphanvu1** [comment]
Thank you. Could you share a compilable project instead of just the inline code? This will make troubleshooting much faster since it'll allow us to reproduce the issue as closely as you do.

**Comment by FVivolo** [comment]
sure, give me some time

**Comment by FVivolo** [comment]
Shoudl be available now in the shared folder

**Comment by FVivolo** [comment]
Sorry to bother you any news or suggestion about this, thanks in advance

**Comment by chauduyphanvu1** [comment]
@FVivolo - I'm getting compilation issues running the sample app. It seems related to how the MoPub SDK is integrated:

```
ld: warning: ignoring file /Users/vchau/Downloads/RuntimeIPhone/MoPubSDKFramework.framework/MoPubSDKFramework, missing required architecture x86_64 in file /Users/vchau/Downloads/RuntimeIPhone/MoPubSDKFramework.framework/MoPubSDKFramework (2 slices)
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_MPMoPubConfiguration", referenced from:
      objc-class-ref in CRunMoPub.o
  "_OBJC_CLASS_$_MPNativeAdRequestTargeting", referenced from:
      objc-class-ref in CRunMoPub.o
  "_OBJC_CLASS_$_MPNativeAdRequest", referenced from:
      objc-class-ref in CRunMoPub.o
  "_OBJC_CLASS_$_MOPUBNativeVideoAdRendererSettings", referenced from:
      objc-class-ref in CRunMoPub.o
  "_OBJC_CLASS_$_MPStaticNativeAdRenderer", referenced from:
      objc-class-ref in CRunMoPub.o
  "_OBJC_CLASS_$_MPInterstitialAdController", referenced from:
      objc-class-ref in CRunMoPub.o
  "_OBJC_CLASS_$_MPRewardedVideo", referenced from:
      objc-class-ref in CRunMoPub.o
  "_kAdMainImageKey", referenced from:
      -[CRunMoPub actLoadNative:] in CRunMoPub.o
  "_OBJC_CLASS_$_MPStaticNativeAdRendererSettings", referenced from:
      objc-class-ref in CRunMoPub.o
  "_kAdIconImageKey", referenced from:
      -[CRunMoPub actLoadNative:] in CRunMoPub.o
      -[CRunMoPub actLoadNativeVideo:] in CRunMoPub.o
  "_OBJC_CLASS_$_MOPUBNativeVideoAdRenderer", referenced from:
      objc-class-ref in CRunMoPub.o
  "_kAdTextKey", referenced from:
      -[CRunMoPub actLoadNative:] in CRunMoPub.o
      -[CRunMoPub actLoadNativeVideo:] in CRunMoPub.o
  "_OBJC_CLASS_$_MPAdView", referenced from:
      objc-class-ref in CRunMoPub.o
  "_OBJC_CLASS_$_MoPub", referenced from:
      objc-class-ref in CRunMoPub.o
  "_kAdTitleKey", referenced from:
      -[CRunMoPub actLoadNative:] in CRunMoPub.o
      -[CRunMoPub actLoadNativeVideo:] in CRunMoPub.o
  "_kAdStarRatingKey", referenced from:
      -[CRunMoPub actLoadNative:] in CRunMoPub.o
      -[CRunMoPub actLoadNativeVideo:] in CRunMoPub.o
  "_MOPUB_BANNER_SIZE", referenced from:
      -[CRunMoPub actAskBanner:] in CRunMoPub.o
  "_kVASTVideoKey", referenced from:
      -[CRunMoPub actLoadNativeVideo:] in CRunMoPub.o
  "_MOPUB_MEDIUM_RECT_SIZE", referenced from:
      -[CRunMoPub actAskBanner:] in CRunMoPub.o
  "_MOPUB_LEADERBOARD_SIZE", referenced from:
      -[CRunMoPub actAskBanner:] in CRunMoPub.o
  "_kAdCTATextKey", referenced from:
      -[CRunMoPub actLoadNative:] in CRunMoPub.o
      -[CRunMoPub actLoadNativeVideo:] in CRunMoPub.o
  "_MOPUB_WIDE_SKYSCRAPER_SIZE", referenced from:
      -[CRunMoPub actAskBanner:] in CRunMoPub.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Are you also seeing that and do you want to send an updated project?

**Comment by chauduyphanvu1** [comment]
I turned on debug log and saw this when rendering the native ad:

```
2019-05-02 14:01:33.094129-0400 RuntimeIPhone[298:9703] 
	[MoPub][MPNativeAdRendererImageHandler] Cell was recycled. Don't bother rendering the image.
2019-05-02 14:01:33.095769-0400 RuntimeIPhone[298:9703] 
	[MoPub][MPNativeAdRendererImageHandler] Cell was recycled. Don't bother rendering the image.
```

It comes from [here](https://github.com/mopub/mopub-ios-sdk/blob/1f4b83e7fefedc9fea517df5448856c63e4f16d9/MoPubSDK/NativeAds/Internal/MPNativeAdRendererImageHandler.m#L45-L48) in the code. The ad view has been recycled and is no longer in the view hierarchy. Could you check if your [superview is not nil](https://github.com/mopub/mopub-ios-sdk/blob/1f4b83e7fefedc9fea517df5448856c63e4f16d9/MoPubSDK/NativeAds/MPStaticNativeAdRenderer.m#L219)?

**Comment by FVivolo** [comment]
I don't get these error here 

i have this set sdkConfig.loggingLevel = MPLogLevelInfo;

sorry don't know what view you are mention, can you explain me more?

does this point to adViewContainer

seems there

![Screen Shot 2019-05-02 at 2 29 04 PM](https://user-images.githubusercontent.com/4428409/57097920-eadada80-6ce6-11e9-808a-2f332e0cdf76.png)

**Comment by chauduyphanvu1** [comment]
If you set `sdkConfig.loggingLevel = MPLogLevelDebug`, which was what I did, you should see additional logging to help you troubleshoot and understand what's going on in the SDK. More info on logging and testing can be found on [this page](https://developers.mopub.com/publishers/ios/test/#enable-logging).

I'm still taking a look at your project, but feel free to use that information to troubleshoot yourself since you are more familiar with the app's setup.

**Comment by FVivolo** [comment]
thanks for the info now I got these errors, here too.

**Comment by FVivolo** [comment]
Hello, wonder if you found any solution to this. thanks in advance for you to try

**Comment by chauduyphanvu1** [comment]
@FVivolo - no progress on this yet but I'll take another look tomorrow! Did you find anything suspicious from your end?

**Comment by FVivolo** [comment]
Hi any progress regarding this issue, did you have an example made using plain UIView and work for you?

**Comment by chauduyphanvu1** [comment]
@FVivolo - we didn't make any new progress since last time. Do you still need help with it? If you've made new changes to your project since then, feel free to send us the updated one.

**Comment by rsatrasala** [comment]
Closing. Please open a new issue if needed.
